File: kaggle_environments/envs/kore_fleets/starter_bots/java/kore/Shipyard.java
package kore;

import java.util.Optional;

public class Shipyard {
    public final String id; 
    public int shipCount; 
    public Point position; 
    public int playerId; 
    public int turnsControlled; 
    public final Board board; 
    public Optional<ShipyardAction> nextAction;
    private final int[] SPAWN_VALUES;
    
    public Shipyard(String shipyardId, int shipCount, Point position, int playerId, int turnsControlled, Board board, Optional<ShipyardAction> nextAction) {
        this.id = shipyardId;
        this.shipCount = shipCount;
        this.position = position;
        this.playerId = playerId;
        this.turnsControlled = turnsControlled;
        this.board = board;
        this.nextAction = nextAction;

        int[] upgradeTimes = new int[9];
        for(int i = 1; i < 10; i++) {
            upgradeTimes[i-1] = (int) Math.pow(i, 2) + 1;
        }
        SPAWN_VALUES = new int[9];
        int current = 0;
        for(int i = 1; i < 10; i++) {
            current += upgradeTimes[i-1];
            SPAWN_VALUES[i-1] = current;
        }
    }

    public Shipyard cloneToBoard(Board board) {
        return new Shipyard(this.id, this.shipCount, this.position, this.playerId, this.turnsControlled, board, this.nextAction);
    }

    public void setNextAction(ShipyardAction action) {
        this.nextAction = Optional.of(action);
    }

    public int maxSpawn() {
        for (int i = 0; i < this.SPAWN_VALUES.length; i++) {
            if (this.turnsControlled < this.SPAWN_VALUES[i]) {
                return i + 1;
            }
        }
        return this.SPAWN_VALUES.length + 1;
    }

    /**
     *  Returns the cell this shipyard is on.
     */
    public Cell cell() {
        return this.board.getCellAtPosition(this.position);
    }

    public Player player() {
        return this.board.players[this.playerId];
    }
   
    /**
     * Converts a shipyard back to the normalized observation subset that constructed it.
     */
    public int[] observation() {
        return new int[]{this.position.toIndex(this.board.configuration.size), this.shipCount, this.turnsControlled};
    }
}

File: kaggle_environments/envs/kore_fleets/starter_bots/ts/kore/Shipyard.ts

import {Board} from "./Board";
import {Cell} from "./Cell";
import {Player} from "./Player";
import {Point} from "./Point";
import {ShipyardAction} from "./ShipyardAction";

const SPAWN_VALUES = [];
const upgradeTimes: number[] = [];
for(let i = 1; i < 10; i++) {
    upgradeTimes[i-1] = Math.pow(i, 2) + 1;
}
let current = 0;
for(let i = 1; i < 10; i++) {
    current += upgradeTimes[i-1];
    SPAWN_VALUES[i-1] = current;
}

export class Shipyard {
    public readonly id: string; 
    public shipCount: number; 
    public position: Point; 
    public playerId: number; 
    public turnsControlled: number; 
    public readonly board: Board; 
    public nextAction: ShipyardAction | undefined;
    
    public constructor(shipyardId: string, shipCount: number, position: Point, playerId: number, turnsControlled: number, board: Board, nextAction: ShipyardAction | undefined) {
        this.id = shipyardId;
        this.shipCount = shipCount;
        this.position = position;
        this.playerId = playerId;
        this.turnsControlled = turnsControlled;
        this.board = board;
        this.nextAction = nextAction;
    }

    public cloneToBoard(board: Board): Shipyard {
        return new Shipyard(this.id, this.shipCount, this.position, this.playerId, this.turnsControlled, board, this.nextAction);
    }

    public setNextAction(action: ShipyardAction): void {
        this.nextAction = action;
    }

    public get maxSpawn(): number {
        for (let i = 0; i < SPAWN_VALUES.length; i++) {
            if (this.turnsControlled < SPAWN_VALUES[i]) {
                return i + 1;
            }
        }
        return SPAWN_VALUES.length + 1;
    }

    /**
     *  Returns the cell this shipyard is on.
     */
    public get cell(): Cell {
        return this.board.getCellAtPosition(this.position);
    }

    public get player(): Player {
        return this.board.players[this.playerId];
    }
   
    /**
     * Converts a shipyard back to the normalized observation subset that constructed it.
     */
    public observation(): number[] {
        return [this.position.toIndex(this.board.configuration.size), this.shipCount, this.turnsControlled];
    }
}